Content.Server/GameObjects/Components/Metabolism/MetabolismComponent.cs

```diff
@@ -0,0 +1,249 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Content.Server.Atmos;
+using Content.Server.GameObjects.Components.Atmos;
+using Content.Server.GameObjects.Components.Body.Circulatory;
+using Content.Shared.Atmos;
+using Content.Shared.Chemistry;
+using Content.Shared.Damage;
+using Content.Shared.GameObjects.Components.Damage;
+using Content.Shared.Interfaces.Chemistry;
+using Robust.Shared.GameObjects;
+using Robust.Shared.IoC;
+using Robust.Shared.Prototypes;
+using Robust.Shared.Serialization;
+using Robust.Shared.ViewVariables;
+
+namespace Content.Server.GameObjects.Components.Metabolism
+{
+    [RegisterComponent]
+    public class MetabolismComponent : Component
+    {
+#pragma warning disable 649
+        [Dependency] private readonly IPrototypeManager _prototypeManager;
+#pragma warning restore 649
+
+        public override string Name => "Metabolism";
+
+        private float _accumulatedFrameTime;
+
+        [ViewVariables(VVAccess.ReadWrite)] private int _suffocationDamage;
+
+        [ViewVariables] public Dictionary<Gas, float> NeedsGases { get; set; }
+
+        [ViewVariables] public Dictionary<Gas, float> ProducesGases { get; set; }
+
+        [ViewVariables] public Dictionary<Gas, float> DeficitGases { get; set; }
+
+        [ViewVariables] public bool Suffocating => SuffocatingPercentage() > 0;
+
+        public override void ExposeData(ObjectSerializer serializer)
+        {
+            base.ExposeData(serializer);
+
+            serializer.DataField(this, b => b.NeedsGases, "needsGases", new Dictionary<Gas, float>());
+            serializer.DataField(this, b => b.ProducesGases, "producesGases", new Dictionary<Gas, float>());
+            serializer.DataField(this, b => b.DeficitGases, "deficitGases", new Dictionary<Gas, float>());
+            serializer.DataField(ref _suffocationDamage, "suffocationDamage", 1);
+        }
+
+        private Dictionary<Gas, float> NeedsAndDeficit(float frameTime)
+        {
+            var needs = new Dictionary<Gas, float>(NeedsGases);
+            foreach (var (gas, amount) in DeficitGases)
+            {
+                var newAmount = (needs.GetValueOrDefault(gas) + amount) * frameTime;
+                needs[gas] = newAmount;
+            }
+
+            return needs;
+        }
+
+        private void ClampDeficit()
+        {
+            var deficitGases = new Dictionary<Gas, float>(DeficitGases);
+
+            foreach (var (gas, deficit) in deficitGases)
+            {
+                if (!NeedsGases.TryGetValue(gas, out var need))
+                {
+                    DeficitGases.Remove(gas);
+                    continue;
+                }
+
+                if (deficit > need)
+                {
+                    DeficitGases[gas] = need;
+                }
+            }
+        }
```

Comment:
Is this going to be expanded to other body needs? should have a more meaningful name if so.